### ISSUE TITLE
How to disable auto focus when quill render?

### ISSUE THREAD

**Issue by kostiantyn-solianyk** [zenoamaro/react-quill#317]
How to disable auto focus when quill render?

**Comment by alexkrolick** [comment]
https://github.com/zenoamaro/react-quill/blob/master/README.md#methods

**Comment by kostiantyn-solianyk** [comment]
I don't know what it is, but when I set a value from my props for example value={this.props.program.description}, then react-quill text field gets focus(). but when I just set value={''"} empty string, react-quill text field works fine and auto focus doesnt appearse

**Comment by FrankYouGithub** [comment]
I have the same problem.When the page is loaded，If there is content in the editor，The page automatically rolls to the editor and auto focus，

**Comment by meiyouqing** [comment]
I have the same issue! Hope someone resolve this.

**Comment by jetzhou** [comment]
Similar issues here, manually calling `blur` on the element doesn't help because the browser has already gotten a `focus` and scrolled to that editor window.

**Comment by alexkrolick** [comment]
Does this change fix your issue: https://github.com/zenoamaro/react-quill/pull/312?

If so I can push a new version

**Comment by jetzhou** [comment]
@alexkrolick I believe the issue is with `dangerouslyPasteHTML` which always grabs the focus afterwards. I created a simple PR that should fix it: https://github.com/zenoamaro/react-quill/pull/321

**Comment by FrankYouGithub** [comment]
@alexkrolick @jetzhou [#321](https://github.com/zenoamaro/react-quill/pull/321/files)can solve this problem

**Comment by codeaid** [comment]
Same issue here.

I have two inputs just above my `ReactQuill` component on the same page. Values of all of them are populated from properties being pushed into the component (as in `value={this.props.something}`). If the Quill editor is empty and I type something into the inputs above the editor suddenly focuses and I end up typing remaining text into it. It doesn't happen if there's content in the editor though.

**Comment by alexkrolick** [comment]
Pushed version 1.2.4

**Comment by kostiantyn-solianyk** [comment]
Version 1.2.4 still not working for me. The page automatically rolls to the editor and auto focus.

**Comment by GEDDE5** [comment]
Same issue on my end — ie. it depends a little on context, but updating `ReactQuill`'s `value` with a prop sometimes grabs undesirable/irremediable focus or else tramples another `element.focus()`.

I experimented with line 68 (https://github.com/zenoamaro/react-quill/pull/321/commits/5a2fc9b5ceeed1befb4c8ff497e04e44ec7f49a8 from @jetzhou's PR,  https://github.com/zenoamaro/react-quill/pull/321) and it works just fine.

**Comment by alexkrolick** [comment]
OK I will check #321 to see what Quill versions support the required clipboard API

**Comment by fabioimpe** [comment]
Same issue here

**Comment by alexkrolick** [comment]
That sounds like a different issue, the component shouldn't be remounting. Check your parent component.

**Comment by fabioimpe** [comment]
#321 resolves my issue...

**Comment by alexkrolick** [comment]
OK, Quill upstream has somewhat OK'd using the `clipboard.convert` method so I pushed 1.2.6 with the fix.

Since it's not a versioned API we'll have to be careful with stability. Please report any issues.

Thanks to everyone who provided input.

**Comment by robertgonzales** [comment]
I'm still seeing this issue on 1.2.6. Quill automatically focuses and scrolls to editor.

**Comment by lukezavrel** [comment]
Auto focus is still occuring in my case too. Any updates on this?

**Comment by vZanchiV** [comment]
I use the latest version to Quill and the problem is not solved ...

my project is in angular 4.

**Comment by iamtoddperkins** [comment]
Too anyone still having this problem (I was).. A much more elegant solution than using window.scroll(0,0) is to simply hide (display:none;) Quill until after its been populated with data. 

I can verify this works in my Vue project, just placing a 'v-show' on the Quill component. I'm sure React has something similar.

**Comment by astegmaier** [comment]
Thanks for the suggestion @iamtoddperkins! I ran across a similar  problem with `formatText` and `removeFormat` always causing Quill to grab focus. The solution was to wrap calls to these methods in `quillContainer.style.display = 'none';` and `quillContainer.style.display = null`. This does seem like something that should just be built into the library, though, or at least documented. I wouldn't expect using the API to modify formats would have the side effect of setting focus - if I wanted that as well, it's easy to just call `quill.focus()`.

**Comment by aipirvu** [comment]
> Too anyone still having this problem (I was).. A much more elegant solution than using window.scroll(0,0) is to simply hide (display:none;) Quill until after its been populated with data.
> 
> I can verify this works in my Vue project, just placing a 'v-show' on the Quill component. I'm sure React has something similar.

This worked for me on an Angular project. Set the [hidden] binding on the editor container and set the binded value to false while I was setting the value and back to true when I finished. Not nice but does the job.

**Comment by guoliucheng** [comment]
May help:
https://blog.csdn.net/nickroprak/article/details/80887808

**Comment by shaayaansayed** [comment]
+1

**Comment by jjalonso** [comment]
Happening to me, all my other inputs that change the props of parent component make react quill to re-render and pull the focus. its imposible to keep typing the other fields.

Any solution for this?

**Comment by botezatumihaicatalin** [comment]
I found out that quill autofocuses the editor when calling 'getFormat' with no parameters.
https://github.com/quilljs/quill/blob/develop/core/quill.js#L271.

Maybe this helps.

**Comment by MarcGodard** [comment]
This is still an issue... why using a hack to get it to work (will try to implement the hack, but why???)

**Comment by cjfradley** [comment]
I'm still having the same issue with my quill editors in a vue project
Various instances on the same page and when setting the v-model content of one of them, the focus is automatically set on that instance which is really annoying
Anyone a fix yet ?

**Comment by alexkrolick** [comment]
Can you raise this on the quilljs github if the issue is upstream? This repo is for the react wrapper

**Comment by cjfradley** [comment]
So I finally found the solution.

AS @alexkrolick pointed out, this is the thread for the react wrapper.
Nevertheless maybe the solution for the vuejs wrapper can help someone:

https://github.com/surmon-china/vue-quill-editor/pull/263

Hope this can help someone

**Comment by jfradley** [comment]
I think this may be a react wrapper issue. I'm having the same issue and it appears to be happening in the `regenerate` logic [here](https://github.com/zenoamaro/react-quill/blob/master/src/component.js#L210).

Maybe we should cache whether it has focus during `regenerate()` and conditionally call `.focus()` at the line linked. 

Should I create another issue for this? I think the original issue is probably solved. This is different and just happens to have the same characteristics.

**Comment by joshuabryant** [comment]
This thread was helpful to me even though I'm using [surmon-china/vue-quill-editor](https://github.com/surmon-china/vue-quill-editor/) and not React.

I landed on another solution that could be useful regardless of your wrapper (particularly if you're stuck with a version that doesn't include the fix), so I thought I'd add it here:

1. Instantiate your Quill editor with the `readOnly` option set to `true`.
2. Then run `quill.enable();` once the editor is ready. (In my case, `vue-quill-editor` fires a `ready` event.)

Maybe less hacky than the `display:none` route?

**Comment by Vanyoo** [comment]
react-quill 1.3.3 still has this problem.

**Comment by imad-mn** [comment]
Confirmed that react-quill 1.3.3 still has this problem.  
I have to inputs (textboxes) in the page,  I cannot write on them and change the state because I immediately go to the quill editor.
In the quill editor, I cannot write more than 1 o 2 characters because it throws me out. I lost focus.

**Comment by aliazizi** [comment]
@alexkrolick the issue still exist in 1.3.5 
this issue should be open

**Comment by avivm** [comment]
For me the issue occurred when I added handlers to the toolbar module.
I fixed it by removing the handlers from the init modules and adding the handler manually afterwards using the toolbar.addHandler method.

**Comment by fancygeek** [comment]
i solve this by pass empty object to the initvalue, not empty string when using ant design  form item

**Comment by ex1t3** [comment]
Still got the issue. I'm using ReactHtmlParser withtin my React_quill ediotr and it autofocuses on state change

**Comment by ex1t3** [comment]
> i solve this by pass empty object to the initvalue, not empty string when using ant design form item

I did this, but it didn't work for me

**Comment by dennmat** [comment]
> I found out that quill autofocuses the editor when calling 'getFormat' with no parameters.

Putting this here to help anyone who it may help. I'm using Vue quill but my issue was I had a dialog that would open and it had an input with a v-model which would cause a re-render of the component when you would type and I was calling `getFormat()` at one point in that re-render. Changing it to:

```
const range = this.editor.getSelection();
getFormat(range.index, range.length);
```

Fixed the issue for me. Thanks @botezatumihaicatalin!

**Comment by taiwanhua** [comment]
Thanks @joshuabryant   's idea

finally, i solved it. you can do like this below :
`
function Editor =(props) =>{

    const [Value, setValue] = useState("");
    const [ReadOnly, setReadOnly] = useState(true);  // for initial readOnly is true , but when Value change then setReadOnly to  false.

    useEffect(() => {
        setValue(props.value);
        if (!isNil(props.value)) {
            setReadOnly(false);
        }
    }, [props.value])

return (
<>
     <ReactQuill
           ref={ref}
           readOnly={ReadOnly}
           theme="snow"
           value={Value || ""}
           ...
      />
 </>
}
`

**Comment by nklaasse** [comment]
When you want to render the editor with a `defaultValue`, the approach down below will also work to prevent autofocusing.

```
useEffect(() => {
    if (quill) {
      // quill?.clipboard.dangerouslyPasteHTML(value); (will result in autofocus)
      quill.root.innerHTML = value;
    }
  }, [quill]);
  ```

**Comment by aleksgrunwald** [comment]
> This thread was helpful to me even though I'm using [surmon-china/vue-quill-editor](https://github.com/surmon-china/vue-quill-editor/) and not React.
> 
> I landed on another solution that could be useful regardless of your wrapper (particularly if you're stuck with a version that doesn't include the fix), so I thought I'd add it here:
> 
> 1. Instantiate your Quill editor with the `readOnly` option set to `true`.
> 2. Then run `quill.enable();` once the editor is ready. (In my case, `vue-quill-editor` fires a `ready` event.)
> 
> Maybe less hacky than the `display:none` route?


This one worked for me.

**Comment by Enigmaticall** [comment]
> 2\. quill.enable();

This worked for me, thank you! Here is a slightly more simplified approach that also fixes defaultValues not working properly:
`const RichTextInputMod = ({ defaultValue, value, setValue, id }) => {
  const [isFirstChange, setIsFirstChange] = useState(true)

  return (
    <ReactQuill
      theme='snow'
      readOnly={isFirstChange}
      value={value}
      onChange={(v) => {
        if (v !== defaultValue && isFirstChange) {
          setIsFirstChange(false)
          return
        }

        setValue(v)
      }}
      id={id}
    />
  )
}`

**Comment by volkorros** [comment]
the same. scrolling to editor :(

**Comment by chaninlaw** [comment]
This way is work for me :)

```tsx
<div
  onClick={() => {
    if (props.readOnly) return
    // Enabled if select inner div
    const editor = quillRef.current?.getEditor() as Quill | undefined
    if (editor) {
      editor.enable()
      editor.focus()
    }
  }}
>
  <QuillStyled
    theme="snow"
    ref={quillRef}
    value={value}
    onChange={handleChange}
    readOnly // set to readOnly
    {...props}
  />
</div>
```

**Comment by lalit-tg** [comment]
For non-react (using quill script)
Turns out - dangerouslyPasteHTML was the culprit when setting default html for the quill, it can be fixed by silently intializing  it.

```
if (hiddenInput?.val()) {
    const delta = quill.clipboard.convert(hiddenInput.val());
    quill.setContents(delta, "silent");
}
```
